# DS-SF-36 | 01 | What is Data Science | Assignment | Starter Code

## Python

> ## "Data scientists' most basic, universal skill is the ability to write code."
>
> Harvard Business Review | “Data Scientists: The Sexiest Job of the 21st Century” (2012)
>
> (https://hbr.org/2012/10/data-scientist-the-sexiest-job-of-the-21st-century)

Today, to give you some more practice beyond the course pre-requisites, we are going to implement a few functions in Python.  This assignment covers to some extend the following topics:

- Functions (defining and using your own functions but also calling functions from packages)
- Loops
- Arithmetic operations
- Conditional statements

**Don't worry if you get stuck.  Ask around, review the answer key, and ask around more.  As this course progresses, your programming proficiency will increase.**

> ### Question 1.  Multiples of 3 and 5
>
> If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9.  The sum of these multiples is 23.
>
> Find the sum of all the multiples of 3 or 5 below 1,000.
>
> (Source: [Project Euler | Problem 1](https://projecteuler.net/problem=1))

In [1]:
# TODO
myList = range(1, 1000)
sum = 0
for num in myList:
    if num%3==0 or num%5==0:
        sum = num + sum
print sum

233168


Answer: 233168

> ### Question 2.  Estimating square roots
>
> Given a real number $m$, let's define the series $u$ as follow:
> - $u_0 = 1$
> - $u_{n+1} = \frac{u_n ^ 2 + m}{2u_n}$
>
>
> Implement the calculations of the series $u$ above to estimate square roots.  Verify that $\sqrt{144} = 12$ and use your function to calculate $\sqrt{1024}$.

In [2]:
# TODO
# Make an array u and initialize the first elem as 0
m = 10000
u = [1]
i = 0
for i in range(0,100):
    next = (u[i]**2 + m)/(2*u[i])
    u.append(next)
print u[100]



100


Answer: TODO

> ### Question 3.  Prime Numbers
>
> A prime (number) is a natural number greater than 1 that has no positive divisors other than 1 and itself.  ([Wikipedia](https://en.wikipedia.org/wiki/Prime_number))
>
> Calculate all primes below 1,000.  What's their sum?

In [3]:
def isPrime(x):
    if x < 2:
        return False
    elif x == 2:
        return True
    else:       
        for n in range(2, int(x**0.5)+1):
            if x % n ==0:
                return False
    return True

def sumList(numList):
    sum = 0
    for el in numList:
        sum = sum + el
    return sum    
                
# TODO
userInput = 1000
primes = []
# Check is each number is a prime and if it is then add to list
for num in range(1, userInput):
    if isPrime(num) == True:
        primes.append(num)    
    
# print primes
# Find the sum of an assay
answer = sumList(primes)
print answer

76127


Answer: 76127

> ### Question 4.  Largest prime factor
>
> The prime factors of 13195 are 5, 7, 13 and 29.
>
> What is the largest prime factor of the number 600851475143?
>
> (Source: [Project Euler | Problem 3](https://projecteuler.net/problem=3))

In [ ]:
# TODO
number = 600851475143
inputList = []

#Check if a number is a prime factor of the input
def primeFactors(number):
    if isPrime(number) == True:
        print "Number is prime"
        answerList = [0]
        return answerList
    else:
        answerList = []
        for num in range(1, int(number ** 0.5) +1):
            if isPrime(num) == True and number % num == 0:
                answerList.append(num)
        print answerList
        return answerList
       
def maxList(inputList):
   answer = max(inputList)
   return answer 
    
inputList = primeFactors(number)
print "prime factors: ", inputList
answer = maxList(inputList)
print "Highest prime factor: ", answer

Answer: TODO

> ### Question 5.  Mean
>
> Write a function to calculate the mean (average) of a list.
>
> What's the mean of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?

In [ ]:
# TODO

Answer: TODO

> ### Question 6.  Sample standard deviation
>
> Write a function to calculate the standard deviation of a sample.
>
> Given the sample $x_1, x_2, ..., x_N$, its standard deviation is defined as $s = \sqrt{\frac{1}{N - 1} \sum_{i = 1}^{N} (x_i  - \bar{x})^2}$, with $\bar{x}$ as the sample mean.
>
> What's the standard deviation of the following sample: 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?
>
> ([Wikipedia](https://en.wikipedia.org/wiki/Standard_deviation#Sample_standard_deviation))

In [ ]:
# TODO

Answer: TODO

> ### Question 7.  Median
>
> Write a function to calculate the median ("middle value") of a list.  ([Wikipedia](https://en.wikipedia.org/wiki/Median))
>
> What's the median of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?

In [ ]:
# TODO

Answer: TODO

> ### Question 8.  Mode
>
> Write a function to calculate the mode ("most frequent value") of a list.  ([Wikipedia](https://en.wikipedia.org/wiki/Mode_(statistics)))
>
> What's the mode of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7 and 5?  How about the mode of 8, 8, 8, 8, 8, 8, 19, 8, 8 and 8?

In [ ]:
# TODO

Answer: TODO